libyul/YulName.cpp

```diff
@@ -0,0 +1,112 @@
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libyul/YulName.h>
+
+#include <libyul/Exceptions.h>
+
+#include <fmt/compile.h>
+
+namespace solidity::yul
+{
+
+YulNameRepository::YulNameRepository()
+{
+	auto const emptyName = defineName("");
+	yulAssert(emptyName == YulNameRepository::emptyName());
+}
+
+std::string_view YulNameRepository::requiredLabelOf(YulName const _name) const
+{
+	auto const label = labelOf(_name);
+	yulAssert(label.has_value(), "YulName currently has no defined label in the YulNameRepository.");
+	return label.value();
+}
+
+std::optional<YulNameRepository::YulName> YulNameRepository::nameOfLabel(std::string_view const _label) const
+{
+	auto const it = std::find(m_definedLabels.begin(), m_definedLabels.end(), _label);
+	if (it != m_definedLabels.end())
+	{
+		YulName labelName{static_cast<YulName>(std::distance(m_definedLabels.begin(), it))};
+		// mostly it'll be iota
+		if (!isDerivedName(labelName) && std::get<0>(m_names[static_cast<size_t>(labelName)]) == labelName)
+			return std::get<0>(m_names[static_cast<size_t>(labelName)]);
+		// if not iota, we have to search
+		auto itName = std::find(m_names.rbegin(), m_names.rend(), std::make_tuple(labelName, YulNameState::DEFINED));
+		if (itName != m_names.rend())
+			return std::get<0>(*itName);
+	}
+	return std::nullopt;
+}
+
+void YulNameRepository::generateLabels(std::set<YulName> const& _usedNames, std::set<std::string> const& _illegal)
+{
+	std::set<std::string> used;
+	std::set<YulName> toDerive;
+	for (auto const name: _usedNames)
+	{
+		if (!isDerivedName(name))
+		{
+			auto const label = labelOf(name);
+			yulAssert(label.has_value());
+			auto const [it, emplaced] = used.emplace(*label);
+			if (!emplaced || _illegal.count(*it) > 0)
+				// there's been a clash ,e.g., by calling generate labels twice;
+				// let's remove this name and derive it instead
+				toDerive.insert(name);
+		}
+	}
+
+	std::vector<std::tuple<std::string, YulName>> generated;
+	auto namesIt = _usedNames.begin();
+	for (size_t nameValue = 1; nameValue < m_names.size(); ++nameValue)
+	{
+		auto name = static_cast<YulName>(nameValue);
+		if (namesIt != _usedNames.end() && name == *namesIt)
+		{
+			if (isDerivedName(name) || toDerive.find(name) != toDerive.end())
+			{
+				std::string const baseLabel(baseLabelOf(name));
+				std::string label (baseLabel);
+				size_t bump = 1;
+				while (used.count(label) > 0 || _illegal.count(label) > 0)
```

Comment:
Some of these names could be more self-explanatory. Especially with `bump`, the name does not help at all until you see how it's actually used :)
```suggestion
				std::string const baseLabel = baseLabelOf(name);
				std::string derivedLabel = baseLabel;
				size_t suffix = 1;
				while (usedLabels.count(derivedLabel) > 0 || _illegaLabels.count(derivedLabel) > 0)
```

Reply:
Also, is `baseLabel` ever expected to be unused? If not it would be clearer to  use `do` ... `while` here.

Reply:
I think that `m_names` may need a better name too, because at this point I still only have a vague idea what it actually stores. But can't suggest a better one until I understand it :)

Reply:
Ah, it stores a dependency to a parent name (if derived) or the index into the vector of actual string labels:)

Reply:
What do you think about replacing the `tuple` with a `struct`? Then you could give this field a name like `labelIndexOrParentName`.

Or we could even make this field a `variant<YulName, string>`. It would actually make searching by label faster (you would not have to find the position of the label in `m_definedLabels` first. It could make the items in `m_names` bigger though (depending on the `string` implementation).

Reply:
Since I complained about the names, here are some more hints on how I'd make them clearer:

- `m_names` -> `m_nameInfo`
    - Similar, but makes it clear that the name is not the value.
- `used` -> `definedUsedNames`
    - I was constantly confusing `used` and `_usedNames`; it's the fact that the ones in `used` are defined that distinguishes them. But TBH the 'used' part is still confusing to me. I'm not sure what it means in this context and if it means the same thing in both names.
- `toDerive` -> `definedUsedNamesToRedefine`
    - It's nice to know from the start that it's a subset of `definedUsedNames`.
- `namesIt` -> `usedNameIt`
    - I kept having to look back remind myself whether it's `name` or `namesIt` that points at `_usedNames`. This info should be always right at hand, in the name.
- `generated` -> `toConvertToDefined` or `toDefine`
- `_illegal` -> `_illegalLabels`

Also for member functions:
- `nameOfLabel()` -> `findNameByLabel()`
    - This is a much heavier operation than I expected when I saw it used. I'd name it like an action rather than as a simple property.